seed/seed.go

```diff
@@ -153,3 +154,75 @@ func ReadSystemEssential(seedDir, label string, essentialTypes []snap.Type, tm t
 
 	return seed20.Model(), seed20.EssentialSnaps(), nil
 }
+
+// ReadSystemEssentialAndBetterEarliestTime retrieves in one go
+// information about the model and essential snaps of the given types
+// for the Core 20 recovery system seed specified by seedDir and label
+// (which cannot be empty).
+// It can operate even if current system time is unreliable by taking
+// a earliestTime lower bound for current time.
+// It returns as well an improved lower bound by considering
+// appropriate assertions in the seed.
+func ReadSystemEssentialAndBetterEarliestTime(seedDir, label string, essentialTypes []snap.Type, earliestTime time.Time, tm timings.Measurer) (*asserts.Model, []*Snap, time.Time, error) {
+	if label == "" {
+		return nil, nil, time.Time{}, fmt.Errorf("system label cannot be empty")
+	}
+	seed20, err := Open(seedDir, label)
+	if err != nil {
+		return nil, nil, time.Time{}, err
+
+	}
+
+	improve := func(a asserts.Assertion) {
+		// we consider only snap-revision and snap-declaration
+		// assertions here as they must be store-signed, see
+		// checkConsistency for each type
+		// other assertions might be signed not by the store
+		// nor the brand so they might be provided by an
+		// attacker, signed using a registered key but
+		// containing unreliable time
+		var tstamp time.Time
+		switch a.Type() {
+		default:
```

Comment:
(nitpick^99) most of the switch/case I have seen has the default as the last entry but that is really probably my most nitpicky comment yet.